src/theory/builtin/theory_builtin_rewriter.cpp

```diff
@@ -330,28 +393,39 @@ Node TheoryBuiltinRewriter::getArrayRepresentationForLambdaRec(TNode n,
     // we will now process the remainder
     curr = next;
     ck = curr.getKind();
+    Trace("builtin-rewrite-debug2")
+        << "  process remainder : " << curr << std::endl;
   }
   if( !rec_bvl.isNull() ){
-    curr = NodeManager::currentNM()->mkNode( kind::LAMBDA, rec_bvl, curr );
+    curr = nm->mkNode(kind::LAMBDA, rec_bvl, curr);
+    Trace("builtin-rewrite-debug") << push;
+    Trace("builtin-rewrite-debug2") << push;
     curr = getArrayRepresentationForLambdaRec(curr, retType);
+    Trace("builtin-rewrite-debug") << pop;
+    Trace("builtin-rewrite-debug2") << pop;
   }
   if( !curr.isNull() && curr.isConst() ){
     // compute the return type
     TypeNode array_type = retType;
-    for( unsigned i=0; i<n[0].getNumChildren(); i++ ){
-      unsigned index = (n[0].getNumChildren()-1)-i;
-      array_type = NodeManager::currentNM()->mkArrayType( n[0][index].getType(), array_type );
+    for (unsigned i = 0; i < size; i++)
+    {
+      unsigned index = (size - 1) - i;
+      array_type = nm->mkArrayType(n[0][index].getType(), array_type);
     }
     Trace("builtin-rewrite-debug2") << "  make array store all " << curr.getType() << " annotated : " << array_type << std::endl;
     Assert(curr.getType().isSubtypeOf(array_type.getArrayConstituentType()));
-    curr = NodeManager::currentNM()->mkConst(ArrayStoreAll(((ArrayType)array_type.toType()), curr.toExpr()));
+    curr = nm->mkConst(ArrayStoreAll(
+        (static_cast<ArrayType>(array_type.toType())), curr.toExpr()));
```

Comment:
This should actually be `ArrayType(array_type.toType())` (it's a construction of an ArrayType, not a cast).